File: src/test/scala/cov_exp_test.scala
package me.zhihan.jacoco.internal

import org.scalatest.FunSuite
import scala.collection.mutable.ArrayBuffer

class CovExpTest extends FunSuite {
  test("Create branch exp") {
    val a = ProbeExp(1)
    val expected = BranchExp(ArrayBuffer(a))
    assert(expected === a.branchExp)

    val b = expected
    val expected2 = BranchExp(ArrayBuffer(b))
    assert(expected2 === b.branchExp)
  }

  test("BranchExp actions") {
    val a = ProbeExp(1)
    val b = a.branchExp
    val c = ProbeExp(2)
    val expected = BranchExp(ArrayBuffer(a, c))

    val id = b.append(c)
    assert(expected === b && id == 1)

    val d = ProbeExp(3)
    b.update(0, d)
    val expected2 = BranchExp(ArrayBuffer(d, c))
    assert(expected2 == b)

    b.join(expected)
    assert(b === BranchExp(ArrayBuffer(d, c, a, c)))
  }

  test("Construction") {
    val a = ProbeExp(1)
    val b = a.branchExp
    val c = ProbeExp(2)
    val expected = BranchExp(ArrayBuffer(a, c))

    val id = b.append(c)
    assert(expected === b && id == 1)

    val d = ProbeExp(3)
    b.update(0, d)
    val expected2 = BranchExp(ArrayBuffer(d, c))
    assert(expected2 == b)

    b.join(expected)
    assert(b === BranchExp(ArrayBuffer(d, c, a, c)))
  }

  test("Eval") {
    val a = ProbeExp(1)
    val b = a.branchExp
    val c = ProbeExp(2)
    val combine = BranchExp(ArrayBuffer(a, c))
    val probes = Array(false, true, false)
    assert(CovExp.evaluate (probes) (a) === true)
    assert(CovExp.evaluate (probes) (c) === false)

    assert(CovExp.evaluate (probes) (combine) == true)
  }

}

File: src/test/java/CovExpTest.java
package org.jacoco.extra.internal;

import static com.google.common.truth.Truth.assertThat;

import org.junit.Test;
import org.junit.Before;
import org.junit.Assert;

public class CovExpTest {
  @Test
  public void testBranchExp() {
    CovExp a = new ProbeExp(0);
    BranchExp actual = a.branchExp();
    assertThat(actual.getBranches()).containsExactly(a);

    BranchExp nested = actual.branchExp();
    assertThat(nested.getBranches()).containsExactly(actual);
  }

  @Test
  public void testAdd() {
    CovExp a = new ProbeExp(0);
    CovExp c = new ProbeExp(1);

    BranchExp b = a.branchExp();
    int idx = b.add(c);

    assertThat(idx).isEqualTo(1);
    assertThat(b.getBranches()).containsExactly(a, c);
  }

  @Test
  public void testUpdate() {
    CovExp a = new ProbeExp(0);
    CovExp c = new ProbeExp(1);
    BranchExp b = a.branchExp();

    b.update(0, c);
    assertThat(b.getBranches()).containsExactly(c);
  }

  @Test
  public void testMerge() {
    CovExp a = new ProbeExp(0);
    BranchExp b = a.branchExp();
    CovExp c = new ProbeExp(1);
    BranchExp d = c.branchExp();
    b.merge(d);
    assertThat(b.getBranches().size()).isEqualTo(2);
    assertThat(b.getBranches()).containsAllOf(a, c);
  }

  @Test
  public void testEvaluate() {
    CovExp a = new ProbeExp(0);
    CovExp b = a.branchExp();
    CovExp c = new ProbeExp(1);
    BranchExp d = a.branchExp();
    d.add(c);

    boolean[] probes = {true, false};

    assertThat(a.eval(probes)).isEqualTo(true);
    assertThat(b.eval(probes)).isEqualTo(true);
    assertThat(c.eval(probes)).isEqualTo(false);
    assertThat(d.eval(probes)).isEqualTo(true);
  }
}
